### ISSUE TITLE
 Complex recurrent layers produce NaN as grad 

### ISSUE THREAD

**Issue by funbotan** [pytorch/pytorch#64186]
## 🐛 Bug

I am trying to run the most basic single-layer RNN with complex inputs on the latest version of PyTorch (1.9.0-CPU). The problem is that the gradient always evaluates to NaN. I’ve tried all recurrent layers (RNN, GRU, LTSM) with the same result.

## To Reproduce

Here is the model:

```
class CRNN(torch.nn.Module):
    def __init__(
        self,
        input_dim: int,
        output_dim: int,
        **kwargs
    ):
        super().__init__(**kwargs)
        self.model = torch.nn.RNN(input_dim, output_dim,
                                  batch_first=True,
                                  dtype=torch.cfloat)
        self.loss = torch.nn.L1Loss()
        self.optimizer = torch.optim.Adam(self.parameters())

    def forward(self, x):
        x = torch.unsqueeze(x, 0)
        return self.model(x)[0]

    def fit(self, x, y):
        self.optimizer.zero_grad()
        z = torch.squeeze(self(x), 0)
        loss = self.loss(z, y)
        loss.backward()
        self.optimizer.step()
        return loss.item()
```

I used `torch.autograd.set_detect_anomaly(True)` and it gave the following results:
- For RNN, the first NaN appears in `AddmmBackward`
- For GRU and LSTM, the first NaN appears in `L1LossBackward`

## Expected behavior

The gradients should be finite

## Environment

 - PyTorch Version (e.g., 1.0): 1.9.0
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.8
 - GPU models and configuration: Currently running on CPU

**Comment by ptrblck** [comment]
Cross-post from [here](https://discuss.pytorch.org/t/complex-recurrent-layers-produce-nan-as-grad/130652) with additional information.

**Comment by anjali411** [comment]
@ptrblck looks like you looked into this before. Is this expected behavior and is it also generically reproducible for other dtypes too?
